### PR TITLE
Challenge UI

### LINKED ISSUE #260: Implement enrolled challenges in the user section of the UI

### LINKED ISSUE #261: Implement challenges in friends section of the UI (+ join button)

### PR DESCRIPTION
**Pull request by bgabriel998**
## Add enrolled challenge list to the user profile 
A challenge Icon has been added next to the collection icon.
After a press on the icon, the challenges that the user are enrolled in are displayed.
For each challenge item, the profile picture of the owner is displayed as well as its name. 

Challenge that have not started yet are display with a dark green background. 

When one press on active challenge, the podium is displayed (or the two first places only, if the number of enrolled users is 2)

A stopwatch-like timer is display to show the challenge remaining time. 

To add a new challenge, a + button has been added on the top right corner above the list of challenges. When the button is clicked, the user must input the name of the challenge (between 3 and 25 chars) and he must also choose a duration (using a list of predefined durations). 

A cup is displayed if the logged in user is currently winning the challenge (see screenshot bellow)

## Preview
image

## Add enrolled challenge list to the friend profile 
image

Close #260 
Close #261

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/PointsChallenge.java [hunk 1]
```diff
@@ -14,31 +19,44 @@
...
+     * @param challengeRanking actual challenge ranking
      */
-    public PointsChallenge(String id, List<String> users, long awardPoints, int status,
+    public PointsChallenge(String id, String founderID, Uri founderUri,
+                           String challengeName, List<String> users, int status,
                            LocalDateTime creationDateTime, int durationInDays,
-                            LocalDateTime startDateTime, LocalDateTime finishDateTime) {
+                           LocalDateTime startDateTime, LocalDateTime finishDateTime,
+                           @Nullable HashMap<String, Integer> challengeRanking,
+                           @Nullable HashMap<String,String> userIDUsername) {
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/PointsChallenge.java:47]: Some parameters are missing in the doc.

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java [hunk 2]
```diff
@@ -30,13 +32,20 @@
...
+     * @param userIDUsernameMap map between user ID and user names
      */
-    public RemotePointsChallenge(String id, List<String> users, long awardPoints, int status,
+    public RemotePointsChallenge(String id, String founderID, Uri founderUri , String challengeName, List<String> users, int status,
                                  LocalDateTime creationDateTime, int durationInDays,
-                                 LocalDateTime startDateTime, LocalDateTime finishDateTime) {
-        super(id, users, awardPoints, status, creationDateTime, durationInDays, startDateTime, finishDateTime);
+                                 LocalDateTime startDateTime, LocalDateTime finishDateTime,
+                                 @Nullable HashMap<String,Integer> userIDPointsMap,
+                                 @Nullable HashMap<String,String> userIDUsernameMap) {
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java:45]: Same as in the last comment (for example, _status_).

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java [hunk 3]
```diff
@@ -45,7 +54,7 @@ public RemotePointsChallenge(String id, List<String> users, long awardPoints, in
      * @param durationInDays duration of the challenge in number of days
      */
     @SuppressLint("NewApi")
-    public static Challenge generateNewChallenge(String founderID, int durationInDays) {
+    public static Challenge generateNewChallenge(String founderID, String challengeName, int durationInDays) {
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java:57]: The new parameter is missing in the doc.

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java [hunk 4]
```diff
@@ -121,6 +127,7 @@ public ChallengeOutcome join() {
         return super.join();
     }
 
+
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java:130]:
```suggestion
```
Added empty line.

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java [hunk 5]
```diff
@@ -193,26 +205,114 @@ private void fillListView() {
...
+        findViewById(R.id.add_challenge).setVisibility(View.VISIBLE);
+        ((TextView)findViewById(R.id.profile_empty_text)).setText(R.string.empty_collection);
+
+        ArrayList<NewChallengeItem> items = new ArrayList<>();
+        List<RemotePointsChallenge> challengeList  =
+                AuthService.getInstance().getAuthAccount().getChallenges().stream().map( c -> (RemotePointsChallenge) c).collect(Collectors.toList());
+
+        for(RemotePointsChallenge enrolledChallenge: challengeList) {
+            NewChallengeItem newChallengeItem;
+            //Retrieve challenge owner name
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java:223]:
```suggestion
            //Retrieve challenge name
```

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java [hunk 6]
```diff
@@ -193,26 +205,114 @@ private void fillListView() {
...
+     * Fill list with challenges the user is enrolled in
+     */
+    private void fillChallengeListView() {
+        // Show correct text if empty
+        findViewById(R.id.add_challenge).setVisibility(View.VISIBLE);
+        ((TextView)findViewById(R.id.profile_empty_text)).setText(R.string.empty_collection);
+
+        ArrayList<NewChallengeItem> items = new ArrayList<>();
+        List<RemotePointsChallenge> challengeList  =
+                AuthService.getInstance().getAuthAccount().getChallenges().stream().map( c -> (RemotePointsChallenge) c).collect(Collectors.toList());
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java:219]:
```suggestion
                displayedAccount.getChallenges().stream().map( c -> (RemotePointsChallenge) c).collect(Collectors.toList());
```
This way this code will work also for the profile of non authenticated users (look at `fillCollectedListView()` for more details).

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     *  @param name challenge name (name of the owner + challenge)
+     * @param numberOfParticipants number of enrolled users
+     * @param startDateTime start time
+     * @param endDateTime end time
+     */
+    public NewChallengeItem(String name, String founderID, Uri founderUri, int status, int numberOfParticipants,
+                            @Nullable LocalDateTime startDateTime, @Nullable LocalDateTime endDateTime
+                            , @Nullable HashMap<String, Integer> challengeRanking,
+                            @Nullable HashMap<String,String> enrolledUsers) {
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java:33]: Some parameters are missing (I suggest we either put or remove all the parameters descriptions). 
However, I find quite confusing the `name` description. Does it represent the name of the challenge or the name of the owner?
If I call `getName()` on a challenge item, I would expect to receive the name of the challenge.

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        return startDateTime;
+    }
+
+    public LocalDateTime getEndDateTime() {
+        return endDateTime;
+    }
+
+    public HashMap<String, Integer> getChallengeRanking(){return challengeRanking;}
+
+    public HashMap<String, String> getEnrolledUsers(){return enrolledUsers;}
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java:71]: Missing doc.

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/Challenge.java [hunk 9]
```diff
@@ -66,15 +80,18 @@
...
     /**
-     * Change challenge status
+     * @return challenge Name
      */
-    void setStatus(ChallengeStatus challengeStatus);
+    String getChallengeName();
 
+    String getFounderID();
+
+    Uri getFounderUri();
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/challenge/Challenge.java:89]: Missing doc.

File: app/src/main/java/ch/epfl/sdp/peakar/general/MainMenuActivity.java [hunk 10]
```diff
@@ -28,6 +29,8 @@
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main_menu);
+
+
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/general/MainMenuActivity.java:33]:
```suggestion
```
Added empty lines.

**Review by giommok — CHANGES_REQUESTED** [review]
Good job with the new challenges, they look really cool!

Check the comments I left, please.

Also, I'm not sure of the choice of the icons in the profile activity.
The collection and challenge icons should be quite similar as they are also really close. Can't we find some icons with the same "style"?
I suggest you ask Alexander (at least, I think it was him) where he took the collection icon from so we can use an icon from the same source.

**Review by ravifrancesco — APPROVED** [review]
Good job, didn't find any issue other than the ones mentioned by @giommok

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java (hunk 3 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java:57]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java (hunk 2 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java:45]: Resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java (hunk 4 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/RemotePointsChallenge.java:130]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java (hunk 6 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java (hunk 7 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java]: Refactored

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java (hunk 8 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/NewChallengeItem.java]: Changed

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/PointsChallenge.java (hunk 1 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/goal/PointsChallenge.java]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/challenge/Challenge.java (hunk 9 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/challenge/Challenge.java]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/general/MainMenuActivity.java (hunk 10 above)
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/general/MainMenuActivity.java]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java [hunk 11]
```diff
@@ -342,6 +442,95 @@ private void updateProfileImage() {
...
+     */
+    public void addChallengeButton(View view){
+        AlertDialog.Builder builder = new AlertDialog.Builder(this);
+        builder.setTitle(getResources().getString(R.string.enter_challenge_name));
+
+        // Set up the input
+        final EditText challengeNameInput = new EditText(this);
+        challengeNameInput.getId();
+
+        // Specify the type of input expected; this, for example, sets the input as a password, and will mask the text
```
[the remark this answers is not in the document]
Comment (bgabriel998) [reply app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java:455]: resolved

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java [hunk 12]
```diff
@@ -194,26 +206,96 @@ private void fillListView() {
         collectionListView.setOnItemClickListener(collectionClicked);
     }
 
+    /**
+     * Setup the challenge list view.
+     * Fill list with challenges the user is enrolled in
+     */
+    private void fillChallengeListView() {
+        // Show correct text if empty
+        if(!(AuthService.getInstance().getAuthAccount() == null)){
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java:215]:
```suggestion
        if(AuthService.getInstance().getAuthAccount() != null){
```

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java [hunk 13]
```diff
@@ -194,26 +206,96 @@ private void fillListView() {
...
+        List<RemotePointsChallenge> challengeList  =
+                displayedAccount.getChallenges().stream().map( c -> (RemotePointsChallenge) c).collect(Collectors.toList());
+
+        for(RemotePointsChallenge enrolledChallenge: challengeList) {
+            NewChallengeItem newChallengeItem;
+
+            newChallengeItem = new NewChallengeItem(
+                    enrolledChallenge,
+                    enrolledChallenge.getFounderID().equals(otherId),
+                    !(AuthService.getInstance().getAuthAccount() == null)
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java:232]:
```suggestion
                    AuthService.getInstance().getAuthAccount() != null
```

File: app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java [hunk 14]
```diff
@@ -50,7 +59,7 @@
 
     private boolean isAuthProfile;
     private Account displayedAccount = null;
-
+    private String otherId = null;
```
Comment (giommok) [inline app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java:62]: This attribute is redundant as the id of other users can be retrieved by doing so:
`((OtherAccount)displayedAccount).getUserID()`
Should we leave it anyway for using the ID more easily?

**Review by giommok — CHANGES_REQUESTED** [review]
Good job again.
I left three small comments.

**Review by giommok — APPROVED** [review]
All good!

**Pull request merged**